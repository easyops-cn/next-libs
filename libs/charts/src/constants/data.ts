export enum DataStatus {
  UnknownError = -1,
  Normal = 0,
  ApiError = 1,
  Empty = 2,
  Forbidden = 3
}
