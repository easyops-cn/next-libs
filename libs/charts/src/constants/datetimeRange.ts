import moment from "moment";

export enum DatetimeRange {
  Latest15Minutes = "now-15m",
  Latest1Hour = "now-1h",
  Latest24Hours = "now-1d",
  Today = "now/d",
  Latest7days = "now-7d",
  Latest30days = "now-30d"
}

export const datetimeRangeList = [
  {
    label: "近15分钟",
    from: DatetimeRange.Latest15Minutes
  },
  {
    label: "近1小时",
    from: DatetimeRange.Latest1Hour
  },
  {
    label: "近24小时",
    from: DatetimeRange.Latest24Hours
  },
  {
    label: "今天",
    from: DatetimeRange.Today
  },
  {
    label: "近7天",
    from: DatetimeRange.Latest7days
  },
  {
    label: "近30天",
    from: DatetimeRange.Latest30days
  }
];

export const timeFilterConditions: { [key: string]: () => string } = {
  [DatetimeRange.Latest15Minutes]: () => "time >= now() - 15m",
  [DatetimeRange.Latest1Hour]: () => "time >= now() - 1h",
  [DatetimeRange.Latest24Hours]: () => "time >= now() - 1d",
  [DatetimeRange.Today]: () =>
    `time >= ${moment()
      .startOf("day")
      .unix()}s AND time <= now()`,
  [DatetimeRange.Latest7days]: () => "time >= now() - 7d",
  [DatetimeRange.Latest30days]: () => "time >= now() - 30d"
};
