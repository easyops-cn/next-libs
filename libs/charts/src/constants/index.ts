export * from "./format";
export * from "./data";
export * from "./datetimeRange";
export * from "./panel";
