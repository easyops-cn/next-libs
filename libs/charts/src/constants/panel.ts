export enum PanelStatus {
  Loading = "loading",
  Normal = "normal",
  PropertyError = "property_error",
  ApiError = "api_error",
  Unknown = "unknown"
}
