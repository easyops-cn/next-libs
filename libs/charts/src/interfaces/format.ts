export interface FormatUnit {
  id: string;
  divisor: number;
  display: string;
}
