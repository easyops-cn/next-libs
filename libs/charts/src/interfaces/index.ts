export * from "./format";
export * from "./panel";
