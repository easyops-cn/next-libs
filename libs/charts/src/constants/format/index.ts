import { percentFormatUnitIds } from "@libs/constants";
import { timeFormatUnitIds } from "./time";
import { dataFormatUnitIds } from "./data";
import { dataRateFormatUnitIds } from "./dataRate";

export enum FormatType {
  None = "none",
  Short = "short",
  Percent = "percent",
  Time = "time",
  Data = "data",
  DataRate = "data_rate"
}

export const formatUnitIds: { [type: string]: string[] } = {
  [FormatType.None]: [],
  [FormatType.Short]: [],
  [FormatType.Percent]: percentFormatUnitIds,
  [FormatType.Time]: timeFormatUnitIds,
  [FormatType.Data]: dataFormatUnitIds,
  [FormatType.DataRate]: dataRateFormatUnitIds
};

export * from "./time";
export * from "./data";
export * from "./dataRate";
