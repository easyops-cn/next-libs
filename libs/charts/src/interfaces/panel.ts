import { FormatType } from "../constants/format";

export interface Grid {
  row?: number;
  column?: number;
}

export interface Fields {
  value: string;
}

export interface Format {
  type?: FormatType;
  precision?: number;
  unit?: string;
}
