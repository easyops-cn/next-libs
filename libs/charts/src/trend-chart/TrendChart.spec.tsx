import React from "react";
import echarts from "echarts";
import { render } from "@testing-library/react";

import {
  formatLabel,
  formatAxisLabel,
  formatMarkLineLabel,
  formatTooltip,
  TrendChartData,
  TrendChart
} from "./TrendChart";

import { Format } from "../interfaces/panel";

import { FormatType, DataRateFormatDisplay } from "../constants/format";

jest.mock("echarts");

describe("TrendChart", () => {
  beforeEach(() => {
    jest.spyOn(echarts, "init").mockReturnValue({
      setOption: jest.fn(),
      resize: jest.fn(),
      dispose: jest.fn(),
      on: jest.fn()
    } as any);
  });

  it("should work", () => {
    const title = "进程内存";
    const data: TrendChartData = {
      legendList: [
        "进程占用内存 (10.0.0.5, 8079)",
        "进程占用内存 (10.0.0.6, 8079)"
      ],
      timeSeries: [1568976541, 1568976602],
      trendDataList: [
        {
          id: "进程占用内存 (10.0.0.5, 8079)",
          name: "进程占用内存 (10.0.0.5, 8079)",
          data: [
            [new Date(1568976541 * 1000), 1000],
            [new Date(1568976602 * 1000), 1000]
          ]
        },
        {
          id: "进程占用内存 (10.0.0.6, 8079)",
          name: "进程占用内存 (10.0.0.6, 8079)",
          data: [
            [new Date(1568976541 * 1000), 1200],
            [new Date(1568976602 * 1000), 1200]
          ]
        }
      ]
    };

    const result = render(
      <TrendChart
        title={title}
        data={data}
        option={{ type: "line" }}
        clickHandler={jest.fn}
      />
    );
    const asFragment = result.asFragment;
    expect(asFragment()).toMatchSnapshot();
  });

  it("should work with data of data rate format", () => {
    const title = "流量趋势";
    const data: TrendChartData = {
      legendList: ["host.net.bits_out", "host.net.bits_out"],
      timeSeries: [1569303000, 1569303060],
      trendDataList: [
        {
          id: "host.net.bits_in",
          name: "host.net.bits_in",
          data: [
            [new Date(1569303000 * 1000), 430],
            [new Date(1569303060 * 1000), 393]
          ]
        },
        {
          id: "host.net.bits_out",
          name: "host.net.bits_out",
          data: [
            [new Date(1569303000 * 1000), 57],
            [new Date(1569303060 * 1000), 57]
          ]
        }
      ]
    };
    const format: Format = {
      type: FormatType.DATA_RATE,
      unit: DataRateFormatDisplay.BYTES_PER_SECOND
    };

    const result = render(
      <TrendChart
        title={title}
        data={data}
        format={format}
        option={{ type: "line" }}
      />
    );
    const asFragment = result.asFragment;
    expect(asFragment()).toMatchSnapshot();
  });

  it("should format label correctlly", () => {
    let label = formatLabel(200, {
      type: FormatType.NONE
    });
    expect(label).toEqual("200");

    label = formatLabel(200, null);
    expect(label).toEqual("200");

    label = formatLabel(0.5, {
      type: FormatType.PERCENT,
      precision: 2
    });
    expect(label).toEqual("50.00%");

    label = formatLabel(0.5, {
      type: FormatType.PERCENT
    });
    expect(label).toEqual("50%");

    label = formatLabel(0, {
      type: FormatType.DATA
    });
    expect(label).toEqual("0");

    label = formatLabel(200, {
      type: FormatType.DATA
    });
    expect(label).toEqual("25 B");

    label = formatLabel(200, {
      type: FormatType.DATA,
      precision: 2
    });
    expect(label).toEqual("25.00 B");

    label = formatLabel(null, {
      type: FormatType.DATA
    });
    expect(label).toEqual(null);

    label = formatLabel(0, {
      type: FormatType.DATA_RATE
    });
    expect(label).toEqual("0");

    label = formatLabel(200, {
      type: FormatType.DATA_RATE
    });
    expect(label).toEqual("25 Bps");

    label = formatLabel(200, {
      type: FormatType.DATA_RATE,
      precision: 2
    });
    expect(label).toEqual("25.00 Bps");

    label = formatLabel(null, {
      type: FormatType.DATA_RATE
    });
    expect(label).toEqual(null);
  });

  it("should format axis label correctlly", () => {
    const label = formatAxisLabel(200, {
      type: FormatType.NONE
    });
    expect(label).toEqual("200");
  });

  it("should format markLine label correctlly", () => {
    const label = formatMarkLineLabel(
      {
        value: 200
      },
      {
        type: FormatType.NONE
      }
    );
    expect(label).toEqual("200");
  });

  it("should format tooltip correctlly", () => {
    let toolTip = formatTooltip(
      {
        type: FormatType.NONE
      },
      [
        {
          seriesName: "host.net.bits_in__sum",
          marker: "<span></span>",
          data: [new Date(1559750400 * 1000), 254]
        }
      ]
    );
    expect(toolTip).toEqual(
      "2019/06/06 0:00:00<br /><span></span>host.net.bits_in__sum: 254<br/>"
    );

    toolTip = formatTooltip(
      {
        type: FormatType.NONE
      },
      [
        {
          seriesName: "host.net.bits_in__sum",
          marker: "<span></span>",
          data: [new Date(1559750400 * 1000), null]
        }
      ]
    );
    expect(toolTip).toEqual(
      "2019/06/06 0:00:00<br /><span></span>host.net.bits_in__sum: 未知<br/>"
    );
  });
});
