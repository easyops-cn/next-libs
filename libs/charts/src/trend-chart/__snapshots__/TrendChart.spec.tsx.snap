// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrendChart should work 1`] = `
<DocumentFragment>
  <div
    class="chart"
    style="position: relative;"
  >
    <div
      class="resize-sensor"
      dir="ltr"
      style="pointer-events: none; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden; max-width: 100%;"
    >
      <div
        class="resize-sensor-expand"
        style="pointer-events: none; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden; max-width: 100%;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; transition: 0s;"
        />
      </div>
      <div
        class="resize-sensor-shrink"
        style="pointer-events: none; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden; max-width: 100%;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; transition: 0s; width: 200%; height: 200%;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TrendChart should work with data of data rate format 1`] = `
<DocumentFragment>
  <div
    class="chart"
    style="position: relative;"
  >
    <div
      class="resize-sensor"
      dir="ltr"
      style="pointer-events: none; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden; max-width: 100%;"
    >
      <div
        class="resize-sensor-expand"
        style="pointer-events: none; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden; max-width: 100%;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; transition: 0s;"
        />
      </div>
      <div
        class="resize-sensor-shrink"
        style="pointer-events: none; position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden; max-width: 100%;"
      >
        <div
          style="position: absolute; left: 0px; top: 0px; transition: 0s; width: 200%; height: 200%;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
